File: java/var-core/src/test/java/com/oselvar/var/core/CanonicalJsonTest.java
package com.oselvar.var.core;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/** Port (concept only — no library) of {@code var-core/tests/canonical-json.test.ts}. */
class CanonicalJsonTest {

    @Test
    void sortsKeysIndentsAndTrailingNewline() {
        var value = Map.of("b", 1, "a", List.of(2, Map.of("d", 4, "c", 3)));
        assertEquals(
                "{\n  \"a\": [\n    2,\n    {\n      \"c\": 3,\n      \"d\": 4\n    }\n  ],\n  \"b\": 1\n}\n",
                CanonicalJson.canonicalStringify(value));
    }

    @Test
    void nonAsciiIsEmittedRaw() {
        var value = Map.of("x", "café 😀");
        assertEquals("{\n  \"x\": \"café 😀\"\n}\n", CanonicalJson.canonicalStringify(value));
    }

    @Test
    void emptyContainersRenderOnOneLine() {
        var value = Map.of("a", List.of(), "b", Map.of());
        assertEquals("{\n  \"a\": [],\n  \"b\": {}\n}\n", CanonicalJson.canonicalStringify(value));
    }

    @Test
    void sortsKeysRegardlessOfInputMapIterationOrder() {
        // Map.of's iteration order is unspecified — build two maps with keys inserted in
        // opposite order and assert both produce the same, key-sorted output.
        var value1 = new java.util.LinkedHashMap<String, Object>();
        value1.put("z", 1);
        value1.put("a", 2);
        value1.put("m", 3);

        var value2 = new java.util.LinkedHashMap<String, Object>();
        value2.put("m", 3);
        value2.put("a", 2);
        value2.put("z", 1);

        String expected = "{\n  \"a\": 2,\n  \"m\": 3,\n  \"z\": 1\n}\n";
        assertEquals(expected, CanonicalJson.canonicalStringify(value1));
        assertEquals(expected, CanonicalJson.canonicalStringify(value2));
    }

    @Test
    void escapesQuotesBackslashesAndControlCharacters() {
        var value = Map.of("s", "a\"b\\c\nd\te");
        assertEquals(
                "{\n  \"s\": \"a\\\"b\\\\c\\nd\\te\"\n}\n", CanonicalJson.canonicalStringify(value));
    }

    @Test
    void serializesNumbersBooleansAndNull() {
        var value = new java.util.LinkedHashMap<String, Object>();
        value.put("int", 1);
        value.put("long", 2L);
        value.put("double", 1.5);
        value.put("bool", true);
        value.put("nul", null);
        assertEquals(
                "{\n  \"bool\": true,\n  \"double\": 1.5,\n  \"int\": 1,\n  \"long\": 2,\n  \"nul\": null\n}\n",
                CanonicalJson.canonicalStringify(value));
    }

    @Test
    void serializesNestedArraysOfObjects() {
        var value = List.of(Map.of("b", 1), Map.of("a", 2));
        assertEquals(
                "[\n  {\n    \"b\": 1\n  },\n  {\n    \"a\": 2\n  }\n]\n",
                CanonicalJson.canonicalStringify(value));
    }

    @Test
    void topLevelScalarSerializesWithoutIndentButWithTrailingNewline() {
        assertEquals("\"hello\"\n", CanonicalJson.canonicalStringify("hello"));
        assertEquals("42\n", CanonicalJson.canonicalStringify(42));
        assertEquals("null\n", CanonicalJson.canonicalStringify(null));
    }
}

File: rust/core/tests/canonical_json_test.rs
//! Port of `CanonicalJsonTest.java` / `canonical-json.test.ts`.

mod common;

use common::vmap;
use varar_core::canonical_json::canonical_stringify;
use varar_core::value::Value;

#[test]
fn sorts_keys_indents_and_trailing_newline() {
    let value = vmap(vec![
        ("b", Value::Int(1)),
        (
            "a",
            Value::list(vec![
                Value::Int(2),
                vmap(vec![("d", Value::Int(4)), ("c", Value::Int(3))]),
            ]),
        ),
    ]);
    assert_eq!(
        "{\n  \"a\": [\n    2,\n    {\n      \"c\": 3,\n      \"d\": 4\n    }\n  ],\n  \"b\": 1\n}\n",
        canonical_stringify(&value)
    );
}

#[test]
fn non_ascii_is_emitted_raw() {
    let value = vmap(vec![("x", Value::from("café 😀"))]);
    assert_eq!("{\n  \"x\": \"café 😀\"\n}\n", canonical_stringify(&value));
}

#[test]
fn empty_containers_render_on_one_line() {
    let value = vmap(vec![("a", Value::list(vec![])), ("b", Value::map(vec![]))]);
    assert_eq!("{\n  \"a\": [],\n  \"b\": {}\n}\n", canonical_stringify(&value));
}

#[test]
fn sorts_keys_regardless_of_input_map_iteration_order() {
    let value1 = vmap(vec![
        ("z", Value::Int(1)),
        ("a", Value::Int(2)),
        ("m", Value::Int(3)),
    ]);
    let value2 = vmap(vec![
        ("m", Value::Int(3)),
        ("a", Value::Int(2)),
        ("z", Value::Int(1)),
    ]);
    let expected = "{\n  \"a\": 2,\n  \"m\": 3,\n  \"z\": 1\n}\n";
    assert_eq!(expected, canonical_stringify(&value1));
    assert_eq!(expected, canonical_stringify(&value2));
}

#[test]
fn escapes_quotes_backslashes_and_control_characters() {
    let value = vmap(vec![("s", Value::from("a\"b\\c\nd\te"))]);
    assert_eq!("{\n  \"s\": \"a\\\"b\\\\c\\nd\\te\"\n}\n", canonical_stringify(&value));
}

#[test]
fn serializes_numbers_booleans_and_null() {
    let value = vmap(vec![
        ("int", Value::Int(1)),
        ("long", Value::Int(2)),
        ("double", Value::Float(1.5)),
        ("bool", Value::Bool(true)),
        ("nul", Value::Null),
    ]);
    assert_eq!(
        "{\n  \"bool\": true,\n  \"double\": 1.5,\n  \"int\": 1,\n  \"long\": 2,\n  \"nul\": null\n}\n",
        canonical_stringify(&value)
    );
}

#[test]
fn serializes_nested_arrays_of_objects() {
    let value = Value::list(vec![
        vmap(vec![("b", Value::Int(1))]),
        vmap(vec![("a", Value::Int(2))]),
    ]);
    assert_eq!(
        "[\n  {\n    \"b\": 1\n  },\n  {\n    \"a\": 2\n  }\n]\n",
        canonical_stringify(&value)
    );
}

#[test]
fn top_level_scalar_serializes_without_indent_but_with_trailing_newline() {
    assert_eq!("\"hello\"\n", canonical_stringify(&Value::from("hello")));
    assert_eq!("42\n", canonical_stringify(&Value::Int(42)));
    assert_eq!("null\n", canonical_stringify(&Value::Null));
}
